python-api/gstswitch/process_monitor.py

```diff
@@ -0,0 +1,157 @@
+
+
+class ProcessMonitor(subprocess.Popen):
+    """Runs a command in a background-thread and monitors its output
+
+    Can block until the command prints a certain string and log the full
+    output into a file
+    """
+
+    def __init__(self, cmd, cmd_output_target=sys.stderr):
+        self.log = logging.getLogger('server-output-monitor')
+
+        # Logfile to write to
+        self._cmd_output_target = cmd_output_target
+
+        # Internal Buffer to search for matched when wait_for_output is called
+        self._buffer = ""
+
+        self.log.debug("starting subprocess")
+
+        try:
+            super(ProcessMonitor, self).__init__(
+                cmd,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                bufsize=0,
+                shell=False)
+        except Exception as err:
+            raise ServerProcessError(err)
+
+        self.log.debug("subprocess successfully started")
+
+    def terminate(self):
+        """Kills the process and waits for the thread to exit"""
+
+        self.log.info("reading remaining data from subprocess")
+        while True:
+            # select takes three lists of file-descriptors to be monitored:
+            #   readable, writeable and exceptional
+            # The last argument is the timeout after which select should give
+            # up. 0 means, that select should return immediately without
+            # blocking (called a "poll")
+            (read, _, _) = select.select([self.stdout], [], [], 0)
+
+            # if the processes' stdout is not readable (ie there is nothing
+            # to read), we're done and can exit the loop
+            if self.stdout not in read:
+                break
+
+            # otherwise read as many bytes as possible, up to 2000,
+            # from the process. os.read does not wait until exactly 2000
+            # bytes have been read but returns as many bytes as possible
+            # the only time os.read block is, when there's nothing to read,
+            # which we ruled out by the poll-call to select before
+            #
+            # os.read -in contrast to self.stdout.read- is non-blocking,
+            # if at least 1 character is readable.
+            self.log.debug("reading data from subprocess")
+            chunk = os.read(self.stdout.fileno(), 2000).decode('utf-8')
+
+            if len(chunk) == 0:
+                break
+
+            self.log.debug(
+                "read %d bytes, appending to cmd_output_target",
+                len(chunk))
+            self._cmd_output_target.write(chunk)
+
+        self.log.debug("terminating the subprocess")
+        super(ProcessMonitor, self).terminate()
+        self.log.debug("waiting for the subprocess to die")
+        super(ProcessMonitor, self).communicate()
+
+    def wait_for_output(self, match, timeout=5, count=1):
+        """Searches the output already captured from the running process for
```

Comment:
Okay, lets get this merged and then look at it again once things are working. I logged https://github.com/timvideos/gst-switch/issues/210 to remind us to do it some point in the future. Good job for some newbie :)